engines/cryomni3d/versailles/engine.cpp

```diff
@@ -0,0 +1,1443 @@
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ *
+ */
+
+#include "common/config-manager.h"
+#include "common/system.h"
+#include "common/error.h"
+#include "common/file.h"
+#include "common/rect.h"
+#include "engines/util.h"
+#include "image/bmp.h"
+
+#include "cryomni3d/font_manager.h"
+#include "cryomni3d/dialogs_manager.h"
+#include "cryomni3d/fixed_image.h"
+#include "cryomni3d/omni3d.h"
+
+#include "cryomni3d/versailles/engine.h"
+
+//#define DEBUG_FAST_START 1
+
+namespace CryOmni3D {
+namespace Versailles {
+
+const FixedImageConfiguration CryOmni3DEngine_Versailles::kFixedImageConfiguration = {
+	45, 223, 243, 238, 226, 198, 136, 145, 99, 113,
+	470
+};
+
+CryOmni3DEngine_Versailles::CryOmni3DEngine_Versailles(OSystem *syst,
+        const CryOmni3DGameDescription *gamedesc) : CryOmni3DEngine(syst, gamedesc),
+	_mainPalette(nullptr), _cursorPalette(nullptr), _transparentPaletteMap(nullptr),
+	_currentPlace(nullptr), _currentWarpImage(nullptr), _fixedImage(nullptr),
+	_transitionAnimateWarp(true), _forceRedrawWarp(false), _forcePaletteUpdate(false),
+	_fadedPalette(false), _loadedSave(-1), _dialogsMan(this),
+	_musicVolumeFactor(1.), _musicCurrentFile(nullptr) {
+}
+
+CryOmni3DEngine_Versailles::~CryOmni3DEngine_Versailles() {
+	delete _currentWarpImage;
+	delete[] _mainPalette;
+	delete[] _cursorPalette;
+	delete[] _transparentPaletteMap;
+
+	delete _fixedImage;
+}
+
+Common::Error CryOmni3DEngine_Versailles::run() {
+	CryOmni3DEngine::run();
+
+	const Common::FSNode gameDataDir(ConfMan.get("path"));
+	SearchMan.addSubDirectoryMatching(gameDataDir, "datas_v/sc_trans", 1);
+	SearchMan.addSubDirectoryMatching(gameDataDir, "datas_v/menu", 1);
+	SearchMan.addSubDirectoryMatching(gameDataDir, "datas_v/basedoc/fonds", 1);
+	SearchMan.addSubDirectoryMatching(gameDataDir, "datas_v/fonts", 1);
+	SearchMan.addSubDirectoryMatching(gameDataDir, "datas_v/spr8col", 1);
+	SearchMan.addSubDirectoryMatching(gameDataDir, "datas_v/spr8col/bmpok", 1);
+	SearchMan.addSubDirectoryMatching(gameDataDir, "datas_v/wam", 1);
+	SearchMan.addSubDirectoryMatching(gameDataDir, "datas_v/objets", 1);
+	SearchMan.addSubDirectoryMatching(gameDataDir, "datas_v/gto", 1);
+	SearchMan.addSubDirectoryMatching(gameDataDir, "datas_v/dial/flc_dial", 1);
+	SearchMan.addSubDirectoryMatching(gameDataDir, "datas_v/dial/voix", 1);
+	SearchMan.addSubDirectoryMatching(gameDataDir, "datas_v/textes", 1);
+	SearchMan.addSubDirectoryMatching(gameDataDir, "datas_v/music", 1);
+	SearchMan.addSubDirectoryMatching(gameDataDir, "datas_v/sound", 1);
```

Comment:
In the Macintosh version I have, on the CDs some of the subdirectories are actually in a `DATAV_HD` directory while others are indeed in `Datas_V`. I just decided to move everything to `Datas_V` when copying to the disk, but if we wanted to support the original structure here are those that are in `DATAV_HD` originally:
- fonts
- GTO
- menu
- spr8col
- textes
- wam

And for information in `DATAV_HD` I also have the following directories that are not listed here (and so far it seems to work without the need to add them to the search paths):
- low_ram
- SAVEGAME

Reply:
Maybe I should remove the datas_v component from the search path. In fact, at the root there is the program and datas_v. This is not really useful, it justs mimics the Windows installation and CDs.
Maybe having every directory directly at root is good enough.

DATAV_HD is the contents of what will be installed on hard drive. There is the same kind of distinction on PC CD: it's still called DATAS_V but placed elsewhere on the disk.
SAVEGAME is used to store saves and it's not need in ScummVM as I use the facilities provided.
I don't know what low_ram contains maybe it's some smaller file to use when there isn't enough memory?

Reply:
The datas_v level could indeed be removed.

On the mac version there is actually no installer. The installation instructions just tell the user to copy a folder from the first CD to the hard drive, and that folder contains both the executable and the DATAV_HD directory. So on the hard drive we end up with a directory that is still named DATAV_HD.

I will check the content of the low_ram directory this evening. I was indeed thinking along the same line as you, but that folder only contains a few txt files from what I remember, so that seemed strange.